R/bambu-processReads_scoreReadClasses.R

```diff
@@ -0,0 +1,212 @@
+#' real gene
+getGeneScore = function(rowData){
+  geneFeatures = prepareGeneModelFeatures(rowData)
+  features = dplyr::select(geneFeatures,!c(labels, names, GENEID))
+  if(checkFeatures(geneFeatures)){
+    geneModel = fit_xgb(as.matrix(features), geneFeatures$labels)
+    geneScore = as.numeric(predict(geneModel, as.matrix(features)))
+    geneFDR = calculateFDR(geneScore, geneFeatures$labels)
+    geneRCMap = match(rowData$GENEID, geneFeatures$names)
+    geneScore = geneScore[geneRCMap]
+    geneFDR = geneFDR[geneRCMap]
+
+  } else {
+    message("Gene Score not calculated")
+    geneScore = rep(1,nrow(rowData))
+    geneFDR = rep(1,nrow(rowData))
+  }
+  return(data.frame(geneScore = geneScore, geneFDR = geneFDR))   
+}
+
+#' calculates the minimum FDR for each score 
+calculateFDR = function(score, labels){
+  scoreOrder = order(score, decreasing = T)
+  orderSave = (1:length(score))[scoreOrder]
+  labels = labels[scoreOrder]
+  score = score[scoreOrder]
+  FDR = cumsum(labels)/(1:length(score))
+  FDR = rev(cummin(rev(FDR)))
+  FDR = FDR[order(orderSave)]
+  return(FDR)
+}
+
+#' calculate and format features by gene for model
+prepareGeneModelFeatures = function(rowData){
+  outData <- as_tibble(rowData) %>% group_by(GENEID) %>% 
+    summarise(numReads = geneReadCount[1],
+              #numReads = sum(readCount, na.rm=T), 
+              names = GENEID[1],
+              labels = !novel[1], 
+              strand_bias = 1-abs(0.5-(sum(readCount.posStrand, na.rm=T)/numReads)), 
+              numRCs=n(), 
+              numExons = max(numExons, na.rm=T), 
+              isSpliced = numExons>1, 
+              # subsets are compatible with 2 or more RCs
+              # or compat with only 1 but are not equal
+              numNonSubsetRCs = numRCs - sum(compatible >= 2 |(compatible == 1 & !equal)),
+              highConfidence=any(confidenceType=='highConfidenceJunctionReads')) %>%
+    mutate(numReads = log2(pmax(1,numReads)))
+  return(outData)
+}
+
+#' ensures that the data is trainable after filtering
+checkFeatures = function(features){
+  labels = features$labels
+  if(sum(labels)==length(labels) | sum(labels)==0){
+    message("Missing presence of both TRUE and FALSE labels.")
+    return(F)
+  }
+  if(length(labels)<50){
+    message("Not enough data points")
+    return(F)
+  }
+  return(T)
+}
+
+#' calculates a score based on how likely a read class is full length
+getTranscriptScore = function(rowData){
+  txFeatures = prepareTranscriptModelFeatures(rowData)
+  features = dplyr::select(txFeatures,!c(labels))
+  if(checkFeatures(txFeatures)){
+    transcriptModel = fit_xgb(features, 
+      txFeatures$labels)
+    txScore = predict(transcriptModel, as.matrix(features))
+
+    #calculates the FDR for filtering RCs based on wanted precision
+    txFDR = calculateFDR(txScore, !rowData$equal)
+
+  } else {
+    message("Transcript Score not calculated")
+    txScore = rep(1,nrow(rowData))
+    txFDR = rep(1,nrow(rowData))
```

Comment:
better set these to NA instead of 1? could be misinterpreted easily when averaging etc